cpp/src/quantiles/groupby.cu

```diff
@@ -83,6 +83,27 @@ namespace cudf {
 
 namespace detail {
 
+gdf_size_type groupby::num_keys() {
+  if (_num_keys > -1)
+    return _num_keys;
+
+  if (has_nulls(_keys)) {
+    auto keys_row_bitmask = row_bitmask(_keys, _stream);
```

Comment:
`keys_row_bitmask` is also used in `keys_sort_order`, so it's probably worthwhile to make computing the `row_bitmask` into a memoized private member function.